homeassistant/components/file_upload/__init__.py

```diff
@@ -145,19 +150,43 @@ async def _upload_file(self, request: web.Request) -> web.Response:
 
         file_upload_data: FileUploadData = hass.data[DOMAIN]
         file_dir = file_upload_data.file_dir(file_id)
-
-        def _sync_work() -> None:
-            file_dir.mkdir()
-
-            # MyPy forgets about the isinstance check because we're in a function scope
-            assert isinstance(file_field, web.FileField)
-
-            with (file_dir / file_field.filename).open("wb") as target_fileobj:
-                shutil.copyfileobj(file_field.file, target_fileobj)
-
-        await hass.async_add_executor_job(_sync_work)
-
-        file_upload_data.files[file_id] = file_field.filename
+        queue: janus.Queue[bytes | None] = janus.Queue()
+
+        def _sync_queue_consumer(
+            sync_q: janus.SyncQueue[bytes | None], _file_name: str
+        ) -> None:
+            file_handle: io.BufferedWriter | None = None
+
+            if file_handle is None:
+                file_dir.mkdir()
+                file_handle = (file_dir / _file_name).open("wb")
```

Comment:
Instead of doing an `open` like this with a finally, use a with statement:

```python
with (file_dir / _file_name).open("wb") as file_handle:
    ...
```

it will automatically close when you're done.